pkg/ingester/label_names_and_values.go

```diff
@@ -161,21 +167,28 @@ func labelValuesCardinality(
 	return nil
 }
 
-func countLabelValueSeries(
-	ctx context.Context,
-	idxReader tsdb.IndexReader,
-	postingsForMatchersFn func(tsdb.IndexPostingsReader, ...*labels.Matcher) (index.Postings, error),
-	lblValMatchers []*labels.Matcher,
-) (uint64, error) {
-	var count uint64
+func countLabelValueSeries(ctx context.Context, lbName string, lbValue string, idxReader tsdb.IndexReader, matchedPostingsCloner *index.PostingsCloner) (uint64, error) {
+	lbPostings, err := idxReader.Postings(lbName, lbValue)
+	if err != nil {
+		return 0, err
+	}
+
+	if matchedPostingsCloner != nil {
+		lbPostings = index.Intersect(lbPostings, matchedPostingsCloner.Clone())
```

Comment:
Food for thought: I think another optimization may be rewrite the `Intersect()` to only return the total number of intersecting postings, and not their list. Something like `index.FindIntersectingPostings()` but returning the count (e.g. `index.CountIntersectingPostings`).

Reply:
If what I'm saying makes sense, then we could open an issue about it.

Reply:
I don't think there's room for an improvement here. `index.Intersect` doesn't return a list, but an iterator. `index.CountIntersectingPostings` would need to copy most of `index.Intersect`.

The only room for an improvement I see here is to do `index.IntersectTwo(a, b index.Postings)` instead of intersecting a variadic number of postings.

Reply:
Actually, `index.IntersectTwo` looks good, but I'd rather send it in a separate PR.

Reply:
I followed up here: https://github.com/grafana/mimir/pull/3071